Frameworks/Accelerate/vImage.cpp

```diff
@@ -951,4 +935,876 @@ vImage_Error vImageMatrixMultiply_ARGB8888(const vImage_Buffer* src,
+    return _vImageUnpremultiplyData_8888<3>(src, dest, flags);
+}
+
+/**
+@Status Interoperable
+@Notes Packed buffer byte order notation is little endian
+*/
+vImage_Error vImageUnpremultiplyData_ARGB8888(const vImage_Buffer* src, const vImage_Buffer* dest, vImage_Flags flags) {
+    return _vImageUnpremultiplyData_8888<0>(src, dest, flags);
+}
+
+/**
+@Status Interoperable
+*/
+vImage_Error vImageBuffer_Init(
+    vImage_Buffer* buffer, vImagePixelCount height, vImagePixelCount width, uint32_t bitsPerFragment, vImage_Flags flags) {
+    const uint32_t truncatedBytesPerFragment = bitsPerFragment >> 3;
+    const uint32_t byteAlignedBytesPerFragment = ((bitsPerFragment + 7) & ~(7)) >> 3;
+    const bool pixelsAreByteAligned = ((bitsPerFragment & 7) == 0) ? true : false;
+    const uint32_t minBitsPerRow = bitsPerFragment * width;
+    const uint32_t minByteAlignedBytesPerRow = ((minBitsPerRow + 7) & ~(7)) >> 3;
+    const uint32_t alignedPitch = _vImageAlignUInt(minByteAlignedBytesPerRow, c_vImageBlockSizeBytes);
+    vImage_Error returnCode = kvImageNoError;
+
+    if ((flags & ~(kvImageNoFlags | kvImageNoAllocate | kvImagePrintDiagnosticsToConsole | kvImageDoNotTile)) != 0) {
+        TraceWarning(TAG, L"Unsupported flag(s)");
+        return kvImageUnknownFlagsBit;
+    }
+
+    FAIL_FAST_IF_FALSE(height > 0 && width > 0 && bitsPerFragment > 0);
+
+    buffer->height = height;
+    buffer->width = width;
+
+    if ((flags & kvImageNoAllocate) == kvImageNoAllocate) {
+        // Align rowBytes
+        buffer->rowBytes = alignedPitch;
+        buffer->data = nullptr;
+        returnCode = c_vImageBlockSizeBytes;
+    } else if (c_vImagePadAllocs == true) {
+        // For 4bytesPerFragment pixels, SSE2 instructions operate on 4 pixels at a time
+        const uint32_t minCompilerAlignmentBytes = 4;
+        const uint32_t additionalRowPaddingBytes = (c_vImageBlockSizeBytes - minCompilerAlignmentBytes);
+        const uint32_t maxPitch = alignedPitch + additionalRowPaddingBytes;
+        const uint32_t allocSize = maxPitch * height;
+
+        // Note: We can't use alignedMalloc here since the client is responsible for freeing the buffer and can use free or aligned_free
+        buffer->data = malloc(allocSize);
+
+        if (buffer->data != nullptr) {
+            // SIMD load/store instructions operate best on 16byte aligned memory and in blocks of 16bytes
+            // Padding to ensure that for unaligned start addresses, at the very least:
+            // 1. A single 16pixel block can be operated on per row
+            // 2. Every second row will be 16byte aligned
+            const uint32_t bytePaddingForIdealAlignment =
+                (c_vImageBlockSizeBytes - ((uint32_t)(buffer->data) & (c_vImageBlockSizeBytes - 1))) % c_vImageBlockSizeBytes;
+            buffer->rowBytes = alignedPitch + bytePaddingForIdealAlignment;
+        } else {
+            // Try again without additional padding
```

Comment:
Also, this isn't the only way a client can create a vImage_Buffer. There is nothing stopping a developer from allocating memory and filling out the data, height, width, rowBytes fields themselves. Therefore, padding can't be assumed and needs to be checked where ever this structure is passed in.